### PR TITLE
Migrate to Oracle DB

### PR DESCRIPTION
**Pull request by SouradeepSaha**
Migrates the database server from Postgres to Oracle DB. Some major changes were needed. Also changes the local database to a docker image of Oracle DB.

### CODE REVIEW

**Comment by haowei99** [comment]
Please change the title of the PR to what the PR actually does - this will help when looking back at previous commits if anything goes wrong

File: CourseParsing/ParseScript.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 def get_course_codes():
     dbc = DatabaseReceiver()
     course_code_query = dbc.select(what="course_codes", table="requirements")
-    course_codes = set([])
+    course_codes = []
```
Comment (haowei99) [inline CourseParsing/ParseScript.py:11]: Is there a reason this is changed from a set to an array? From what I remembered the requirements table has reqs from many years - which has duplicate course codes (for ex MATH 135 in many programs/years) and I think that was why a set was used. 

Have you tested this out?

File: CourseParsing/ParseScript.py [hunk 2]
```diff
@@ -20,7 +20,7 @@ def get_course_codes():
 
 
 if __name__ == "__main__":
-    course_codes = {'PACS', 'MATBUS', 'HLTH', 'ME', 'ARBUS', 'MSCI', 'ACTSC', 'CHE', 'CHEM', 'STAT',
+    course_codes = ['PACS', 'MATBUS', 'HLTH', 'ME', 'ARBUS', 'MSCI', 'ACTSC', 'CHE', 'CHEM', 'STAT',
```
Comment (haowei99) [inline CourseParsing/ParseScript.py:23]: same here with the set/array usage

File: CourseParsing/ParseScript.py [hunk 3]
```diff
@@ -31,15 +31,15 @@ def get_course_codes():
...
+                    'THPERF', 'UNIV', 'VCULT']
 
     dbc = DatabaseSender()
 
-    dbc.execute("DROP TABLE IF EXISTS " + dbc.antireqs_table + ";")
-    dbc.execute("DROP TABLE IF EXISTS " + dbc.prereqs_table + ";")
-    dbc.execute("DROP TABLE IF EXISTS " + dbc.communications_table + ";")
-    dbc.execute("DROP TABLE IF EXISTS " + dbc.breadth_table + ";")
-    dbc.execute("DROP TABLE IF EXISTS " + dbc.course_table + ";")
+    dbc.drop_table(dbc.antireqs_table)
```
Comment (haowei99) [inline CourseParsing/ParseScript.py:38]: If you are a new member to UWPath, will you run into errors since you had no tables to drop to begin with? I think that's why the IF EXISTS statement exist - unless the flow has changed

File: Database/DatabaseConnection.py [hunk 4]
```diff
@@ -7,31 +7,38 @@
...
             password = os.getenv("DB_PASS")
         if os.getenv("DB_USER") is not None:
             user = os.getenv("DB_USER")
-        if os.getenv("DB_NAME") is not None:
-            database = os.getenv("DB_NAME")
-        if os.getenv("DB_HOST") is not None:
-            host = os.getenv("DB_HOST")
-        if os.getenv("DB_PORT") is not None:
-            port = os.getenv("DB_PORT")
+        if os.getenv("ORACLE_DSN") is not None:
```
Comment (haowei99) [inline Database/DatabaseConnection.py:25]: Make sure you have this thing saved somewhere secure if needed to pull it up again :)

File: Database/DatabaseSender.py [hunk 5]
```diff
@@ -19,17 +19,18 @@ def create_requirements(self):
         create requirements table for each major
         """
         command = "CREATE TABLE " + self.requirements_table + """ (
-            id SERIAL PRIMARY KEY,
+            id NUMBER GENERATED BY DEFAULT AS IDENTITY,
```
Comment (haowei99) [inline Database/DatabaseSender.py:22]: what's the difference here?

File: Database/DatabaseSender.py [hunk 6]
```diff
@@ -19,17 +19,18 @@ def create_requirements(self):
...
+            course_codes VARCHAR(4000),
             number_of_courses int,
             credits_required DECIMAL(4,2),
             additional_requirements VARCHAR(255), 
             major_name VARCHAR(255),
             faculty VARCHAR(50),
             link VARCHAR(255),
-            year VARCHAR(20)
+            year VARCHAR(20),
+            PRIMARY KEY (id)
```
Comment (haowei99) [inline Database/DatabaseSender.py:33]: if we have "id SERIAL PRIMARY KEY", we don't need to declare Primary Key here?

What's the reason behind this change?

File: Database/DatabaseSender.py [hunk 7]
```diff
@@ -39,7 +40,7 @@ def create_courses(self):
         Creates courses table.
         """
         # TODO - remove antireq field
-        command = "CREATE TABLE IF NOT EXISTS " + self.course_table + """ (
+        command = "CREATE TABLE " + self.course_table + """ (
```
Comment (haowei99) [inline Database/DatabaseSender.py:43]: similar to my concerns before - what if we had a table already

File: Database/DatabaseSender.py [hunk 8]
```diff
@@ -48,7 +49,7 @@ def create_courses(self):
             credit VARCHAR(255),
             info VARCHAR(2000),
             offering VARCHAR(255),
-            online BOOLEAN,
+            is_online INT,
```
Comment (haowei99) [inline Database/DatabaseSender.py:52]: Why int?

File: Database/DatabaseSender.py [hunk 9]
```diff
@@ -87,25 +88,26 @@ def create_communications(self):
         """
         Creates communications table.
         """
-        command = "CREATE TABLE IF NOT EXISTS " + self.communications_table + """ (
-            id SERIAL PRIMARY KEY,
+        command = "CREATE TABLE " + self.communications_table + """ (
```
Comment (haowei99) [inline Database/DatabaseSender.py:91]: same here as before

File: Database/DatabaseSender.py [hunk 10]
```diff
@@ -140,7 +143,8 @@ def insert_antireqs(self, code, antireqs):
         not_exist += "WHERE course_code = '" + code + "'"
         command = "INSERT INTO " + self.antireqs_table + " (course_code, antireq, extra_info)"
         command += "\nSELECT '" + code + "', '" + antireqs.str("antireqs") + "', '" + antireqs.str("extra") + "'\n"
-        command += "WHERE NOT EXISTS (\n" + not_exist + "\n);"
+        command += "FROM dual\n"
```
Comment (haowei99) [inline Database/DatabaseSender.py:146]: what is dual?

**Review by haowei99 — CHANGES_REQUESTED** [review]
left comments, mostly about if a new flow is established and if edge cases have been tested

File: CourseParsing/ParseScript.py (hunk 1 above)
Comment (SouradeepSaha) [reply CourseParsing/ParseScript.py:11]: Line 50 reverses the course_codes, which is not possible for a set. So we converted it to a list, unless you wanted different functionality.

File: CourseParsing/ParseScript.py (hunk 2 above)
Comment (SouradeepSaha) [reply CourseParsing/ParseScript.py:23]: Same as above. See line 50

File: CourseParsing/ParseScript.py (hunk 3 above)
Comment (William-Gao1) [reply CourseParsing/ParseScript.py:38]: Oracle does not have IF EXISTS. The drop_table function handles it

File: Database/DatabaseSender.py (hunk 5 above)
Comment (William-Gao1) [reply Database/DatabaseSender.py:22]: Oracle does not have SERIAL type

File: Database/DatabaseSender.py (hunk 6 above)
Comment (William-Gao1) [reply Database/DatabaseSender.py:33]: Oracle does not have SERIAL type

File: Database/DatabaseSender.py (hunk 7 above)
Comment (William-Gao1) [reply Database/DatabaseSender.py:43]: See above

File: Database/DatabaseSender.py (hunk 8 above)
Comment (William-Gao1) [reply Database/DatabaseSender.py:52]: Oracle does not support BOOLEAN so it needs to be an int

File: Database/DatabaseSender.py (hunk 9 above)
Comment (William-Gao1) [reply Database/DatabaseSender.py:91]: see above

File: Database/DatabaseSender.py (hunk 10 above)
Comment (William-Gao1) [reply Database/DatabaseSender.py:146]: WHERE clause needs FROM clause. Dual is just a dummy table for this purpose

File: Database/DatabaseConnection.py (hunk 4 above)
Comment (William-Gao1) [reply Database/DatabaseConnection.py:25]: not sure what you mean here. We have the values, but they will be injected when we build the pipeline

File: CourseParsing/ParseScript.py (hunk 3 above)
Comment (William-Gao1) [reply CourseParsing/ParseScript.py:38]: more specifically, the drop_table function defines a db procedure to check if the table exists and if it does it will execute the drop table command. Since we are dropping all tables at the start of each dataparse, no need to check if the table exists when we recreate them as you mention in your comments below

File: CourseParsing/ParseScript.py (hunk 1 above)
Comment (haowei99) [reply CourseParsing/ParseScript.py:11]: Sounds good

Comment (haowei99) [reply CourseParsing/ParseScript.py:11]: I just wanted to make sure no duplicated courses are displayed

File: CourseParsing/ParseScript.py (hunk 3 above)
Comment (haowei99) [reply CourseParsing/ParseScript.py:38]: Ok sounds good so oracle does have a safety check

File: Database/DatabaseConnection.py (hunk 4 above)
Comment (haowei99) [reply Database/DatabaseConnection.py:25]: that sounds good, just make sure the value isn't in the public repo somewhere for security reasons

File: Database/DatabaseSender.py (hunk 10 above)
Comment (haowei99) [reply Database/DatabaseSender.py:146]: would be nice if you can add a comment above the code to explain this - so it is more intuitive to other people reading the code

**Review by haowei99 — APPROVED** [review]

File: CourseParsing/ParseScript.py (hunk 3 above)
Comment (William-Gao1) [reply CourseParsing/ParseScript.py:38]: Not sure what you mean. Oracle has no built in equivalent to IF EXISTS. drop_table is just a function I wrote to manually check if a table exists. Feel free to take a look

File: Database/DatabaseConnection.py (hunk 4 above)
Comment (William-Gao1) [reply Database/DatabaseConnection.py:25]: Of course. Sounds good

File: Database/DatabaseSender.py (hunk 8 above)
Comment (William-Gao1) [reply Database/DatabaseSender.py:52]: Also 'online' is a reserved keyword so I changed it to 'is_online'
